### ISSUE TITLE
Implement rule extensions

### ISSUE THREAD

**Issue by comius** [bazelbuild/bazel#19507]
Checklist:
- [x] extending from parent
- [x] calling parent's implementation function (in review)
- [x] merging public attributes
- [x] intializers in extended rules
- [x] limiting rule's extendability

Design document: https://docs.google.com/document/d/1p6z-shWf9sdqo_ep7dcjZCGvqN5r2jsPkJCqHHgfRp4/edit

**Comment by rbtcollins** [comment]
@comius where do you want bugs for this? We've kicked the tires on it a little and found some friction - our use case is decorating language rule  foo_test to provide external dependencies - and had issues with both getting the parent executable cleanly, but also dealing with _allowlist_function_transition in the extended rule.

**Comment by comius** [comment]
You can open separate bugs or explain more here. I'm aware there might be some problems with executables in executable rules. I'll need to figure this out.

About `_allowlist_function_transition`. The implicit attribute should be added automatically when a Starlark transition is used. Do you have more specific case where this is causing a problem?

**Comment by comius** [comment]
No, it’s unintentional. C++ rules should be extendable in current form already. I’ll have a look.

**Comment by timothyg-stripe** [comment]
@comius I was looking into creating a rule that extends `java_library`, which I thought would be a supported use case given the number of examples in the [design doc](https://docs.google.com/document/d/1p6z-shWf9sdqo_ep7dcjZCGvqN5r2jsPkJCqHHgfRp4/edit) that use `java_library`. Unfortunately, I found that this is not possible (at least on Bazel 7.0.1 with `--experimental_rule_extension_api=true`):

> Error in rule: The rule 'java_library' is not extendable. Only Starlark rules not using deprecated features (like implicit outputs, output to genfiles) may be extended. Special rules like analysis tests or rules using build_settings cannot be extended.

It looks like `java_library` isn't extendable because it has two [implicit outputs](https://cs.opensource.google/bazel/bazel/+/master:src/main/starlark/builtins_bzl/common/java/java_library.bzl;l=381-382;drc=24f118232394a94f6bf0888dc9b5978425edca55). Do you have any plans to support extending rules with implicit outputs, or else remove them entirely from `java_library`? We'd rather not resort to `--experimental_builtins_bzl_path`, as the ability to piggy-back off of native rules w/o maintaining our fork of `@_builtins` is a primary reason why we'd like to use rule extensions in the first place.

**Comment by mering** [comment]
Looks like `--experimental_rule_extension_api` was flipped in Bazel 8 (https://github.com/bazelbuild/bazel/commit/80d8e4253332961b703f67c0b33e6e6f803dd5b7). @comius Can this issue be closed?

**Comment by alexeagle** [comment]
I'm also curious about that, and whether the feature can be marked as no longer experimental?
https://bazel.build/versions/9.0.0/rules/lib/builtins/Subrule is pretty scary for considering using this in a widely-used semver-stable ruleset.
